### ISSUE TITLE
Bounce time of 5 seconds is too short for H35i.

### ISSUE THREAD

**Issue by rainwoodman** [dahlb/ha_blueair#164]
When I click on Led Light switch from on to off, for example, it now first displays off, then flips back to on then flips to off.

**Comment by dahlb** [comment]
yeah I think that's the difference between the sleep it used to be and deferring to ha's choice of 5 seconds which isn't always actually 5 seconds but seems to act more as a max of 5 seconds

**Comment by rainwoodman** [comment]
I see. The debouncer only guarantees no calls within the five seconds of
the first request, but it does not guarantee the call is five seconds after
the last request. We really needed is a debounser that postpones till the
last request is five seconds old. Seems to be a feature request to ha core
but I think we can start with a local fix — not sure how yet. Have you
filed such features to ha core before? How much friction are there?

**Comment by dahlb** [comment]
I wouldn't expect core to implement a waiting feature. I think their goal with the debouncer is to reduce load on api infrastructure not account for a poorly implemented api. if we put the async sleep back in before invoking the update function, being an async sleep it doesn't tie up any resources and accomplishes the need caused by blueair's api implementation limitation.

**Comment by dahlb** [comment]
I added back in the async sleep 5, we can leave this open if you would rather core handle this and link this issue to the one for core

**Comment by rainwoodman** [comment]
I have not read your change (on the phone) but I have been thinking adding a delay to the refresher could keep the other benefit of the bouncer which is improving the UI responsiveness. If the delay is in the setter then it will stop the setter from returning (blocking the calling coroutine even if it is not occupying CPU). If it is in the refresher the bouncer caps it after the setter has returned. It will also ensure no update is performed within five seconds of the last request.

**Comment by dahlb** [comment]
Thanks for helping me see the issue with the sleep, I think this will fix the debouncer https://github.com/dahlb/ha_blueair/commit/f91e4015c634ba1c14737a4e0e9370c844fd37c9 I found a parameter immediate and set it to false which seems to get the behavior we want

**Comment by rainwoodman** [comment]
I thought we had immediate=False and cooldown=10 previously, which worked for my h35i devices when I tested.

The problem started to surface after the cooldown was reduced to 5 seconds I thought. I didn't realize you changed the immediate=True?

immediate=False was the behavior that guarantees the first refresh occurs after 5(cooldown) seconds of the request, and that's not safe for the case we would like to guarantee: a single user changing the state of the device. For that we need to ensure the first refresh occurs after 5 seconds of the last request that were queued into the debouncer. 

Let me poke with moving the delay to the refresh and see if that is an effective workaround.

**Comment by rainwoodman** [comment]
I did some experiments and the problem is not because of our refresh. Apparently it was home assistant flipping the status before even calling the refresh() function we supplied. I also noticed our callback system was never used by home assistant which seems to be unexpected and the root cause of this?

**Comment by rainwoodman** [comment]
I think it is the callback system (or the lack of) e.g. https://developers.home-assistant.io/docs/integration_fetching_data/#push-vs-poll and https://developers.home-assistant.io/docs/core/entity/#subscribing-to-updates

I tried with several bouncer configurations. Home assistant insists flipping the state back to the value recorded in the DB in about 5 seconds, without calling any of our device refresh routines (without talking to blueair server, that is). That explained why if we do a forced state refresh in 5 seconds (which updates the state tracked in HA database) we do not observe the flipping anymore.

Then I added a callback for attributes changes, and the flipping is mostly fixed. However what worries me is that now I get logs like this:

```
.0)
2024-12-15 22:50:00.754 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity light.kitchen_led_light=on>>
2024-12-15 22:50:00.755 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity switch.kitchen_night_mode=off>>
2024-12-15 22:50:00.755 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity binary_sensor.kitchen_water_shortage=off>>
2024-12-15 22:50:00.756 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity binary_sensor.kitchen_online=on>>
2024-12-15 22:50:00.756 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity switch.kitchen_auto_fan_mode=off>>
2024-12-15 22:50:00.756 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity switch.kitchen_wick_dry_mode=off>>
2024-12-15 22:50:00.757 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity fan.kitchen_fan=off>>
2024-12-15 22:50:00.757 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity switch.kitchen_child_lock=off>>
2024-12-15 22:50:00.757 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity binary_sensor.kitchen_filter_expiration=off>>
2024-12-15 22:50:00.758 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity sensor.kitchen_humidity=unavailable>>
2024-12-15 22:50:00.758 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity binary_sensor.kitchen_child_lock=off>>
2024-12-15 22:50:00.758 ERROR (MainThread) [blueair_api.callbacks] Kitchen publishing updates <bound method Entity.schedule_update_ha_state of <entity sensor.kitchen_temperature=unavailable>>

```

aka all entity states in HA db are forcefully updated when any of the state is changed. I think this is most likely not going to be blessed by HA core. We need some kind of per attribute callback if callback is the right way. I will share a draft PR for wiring the callback. But we need some thinking about how to best do the per attribute callback.

**Comment by rainwoodman** [comment]
One way for the per-attribute callback is to add a new string argument for the name of the attribute that actually mutated.
Another way is to upgrade the attributes to objects:

class Attribute[T](CallbackMixin):
   @property
   def value(self)  -> T | None:
      ...

   @property
   def is_implemented(self) -> bool:
      ...
   async def set_value(self, value: T):
      ...  
```

**Comment by dahlb** [comment]
we shouldn't need the callbacks, looking at https://developers.home-assistant.io/docs/integration_fetching_data/#push-vs-poll we should be able to return the dataclass and let the coordinator rely on the __eq__ method

**Comment by rainwoodman** [comment]
That polling mechanism (coordinator based polling) is the one we were using
with 5 / 10 minutes of polling frequency. There is an automated HA state
update after the coordinator refresh indeed but the entity triggered state
change does not occur soon enough for the HA state to catch up: we need to
sync the HA state to the local state cache not the remotely polled state.
Or you mean there is a way to repurpose the coordnator for the entity state
updates?

There is indeed a way to avoid using the callbacks: we can modify the
entity to force trigger a HA state update at the time of state mutation
(right after calling coordinator.set_xxxx). That most likely would mean HA
will pick up the locally cached new state value, and not reverting the
entity 'displayed' state to the old state.

**Comment by dahlb** [comment]
Can you tell me steps to reproduce this flicker?  I haven't been able to see it, also have you looked at the update I made this morning to the coordinator update function based on the article from above, I think that should have solved any flicker based on the design details I read in the article (sorry I can't provide sha links, on my phone atm)

**Comment by rainwoodman** [comment]
I tried on the head of main branch (fd8a018701c4ada1b56b1c318119aa4ee5211824). It was not even updating the state after I clicked on the LED light button. The fan on/off seems to work fine. An earlier commit the LED light button was flickering but was still functional.

https://github.com/user-attachments/assets/1c55cfa9-3825-41d7-a179-4015451b712d

**Comment by rainwoodman** [comment]
Recording at cc60f62f9e7e8f289a863b5d4ef8b8f8644bab85, with pointer. This shows the flickering effect.
https://github.com/user-attachments/assets/0ace6993-ecea-4fe2-842c-0b4dac94052a

**Comment by rainwoodman** [comment]
Mystery solved. I think we are just missing the call to self.async_write_ha_state() in light.py -- that call exists in switch.py and fan.py, but not in light.py. Since apparently other switches and lights were working I think that's the only fix we needed.

The change in fe89b67defb824db48583cd17c5f858855afea53 made bug worse when we are missing that update (because it skipped the entity update at all); but I checked apply the fix on top of fe89b6 the led light switch is no longer flickering.

**Comment by dahlb** [comment]
thanks, makes sense and I missed it because I was checking with fan percentage changes
